### ISSUE TITLE
Usage of custom defined clip instead of numpy.clip ?

### ISSUE THREAD

**Issue by George3d6** [lmcinnes/umap#84]
I've noticed a custom `clip(val)` function that clips between -4 and 4 is used, instead of just using `np.clip(val,-4,4)`. Is this done on purpose (e.g. does it improve performance) or was it just an oversight ?

**Comment by lmcinnes** [comment]
It's in a tight loop and numba does a better job of compiling an
effectively inlining a simple function like that than dealing with the
numpy. Or it did the last time I checked -- numba is always improving so
maybe np.clip now in a robust way. If you want to run some experiments with
a recent version of numba I would be interested to see the results.

**Comment by George3d6** [comment]
Tried doing some very basic  benchamrks:

import numba
import numpy as np
from random import uniform
import time

```

@numba.njit()
def myclip(val):
    """Standard clamping of a value into a fixed range (in this case -4.0 to
    4.0)

    Parameters
    ----------
    val: float
        The value to be clamped.

    Returns
    -------
    The clamped value, now fixed to be in the range -4.0 to 4.0.
    """
    if val > 4.0:
        return 4.0
    elif val < -4.0:
        return -4.0
    else:
        return val
    
nr = pow(10,6) * 3
nrs = np.zeros(nr)
for i in range(0,nr):
    nrs[i] = uniform(-6.5,6.5)
    
def test1():
    for i in range(0,nr):
        val = myclip(nrs[i])
        
@numba.jit()           
def test2():
    for i in range(0,nr):
        val = np.clip(nrs[i],-4,4)

start = time.time()        
test1()
end = time.time()
print('Custom function takes: {}'.format(end - start))

start = time.time()        
test2()
end = time.time()
print('Numpy function takes: {}'.format(end - start))
```

By the looks of it the custom function is indeed 3-5 times faster than numpy.clip. So considering it's called a lot in some nested for loops, the performance gain there might not be trivial.

Will close this.

**Comment by lmcinnes** [comment]
Thanks for verifying this. I seem to recall having done this myself, but
that was a long time ago, and Numba has changed a lot. Good to know that
the current code is still the right way to go for now.
